sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -764,61 +768,202 @@ static void set_codec_init_func(const struct snd_soc_acpi_link_adr *link,
 	}
 }
 
-static int create_sdw_codec_dai(struct device *dev,
-				struct snd_soc_dai_link *dai_links, int *id,
-				struct snd_soc_dai_link_component *cpus,
-				const struct snd_soc_acpi_link_adr *link,
-				int *cpus_id)
+/*
+ * check endpoint status in slaves and gather link ID for all slaves in
+ * the same group to generate different CPU DAI. Now only support
+ * one sdw link with all slaves set with only single group id.
+ *
+ * one slave on one sdw link without aggregaged = 0
+ * one sdw link <---> one-cpu DAI <---> one-codec DAI
+ *
+ * two or more slaves on one sdw link without aggregaged = 0:
+ * one sdw link <---> one-cpu DAI <---> multi-codec DAIs
+ *
+ * multiple links with multiple slaves with aggregated = 1:
+ * link N <---> cpu DAIs N <----> codec DAIs I (I >= 1)
+ * link N <---> cpu DAIs M <----> codec DAIs J (J >= 1)
+ * ...
+ */
+static int get_slave_info(const struct snd_soc_acpi_link_adr *adr_link,
+			  int *cpu_dai_id, int *cpu_dai_num,
+			  int *codec_num, int *group_id,
+			  bool *group_generated)
+{
+	const struct snd_soc_acpi_adr_device *adr_d;
+	const struct snd_soc_acpi_link_adr *adr_next;
+	int index = 0;
+
+	*codec_num = adr_link->num_adr;
+	adr_d = adr_link->adr_d;
+
+	/* mask indicates only one sdw link */
+	if (!is_power_of_2(adr_link->mask))
+		return -EINVAL;
+
+	cpu_dai_id[index++] = ffs(adr_link->mask) - 1;
+	if (!adr_d->endpoints->aggregated) {
+		*cpu_dai_num = 1;
+		*group_id = 0;
+		return 0;
+	}
+
+	*group_id = adr_d->endpoints->group_id;
+
+	/* gather other link ID of slaves in the same group */
+	for (adr_next = adr_link + 1; adr_next && adr_next->num_adr;
+		adr_next++) {
+		const struct snd_soc_acpi_endpoint *endpoints;
+
+		endpoints = adr_next->adr_d->endpoints;
+		if (!endpoints->aggregated ||
+		    endpoints->group_id != *group_id)
+			continue;
+
+		if (!is_power_of_2(adr_next->mask))
+			return -EINVAL;
+
+		cpu_dai_id[index++] = ffs(adr_next->mask) - 1;
+		*codec_num += adr_next->num_adr;
+	}
+
+	/*
+	 * indicate CPU DAIs for this group have been generated
+	 * to avoid generating CPU DAIs for this group again.
+	 */
+	group_generated[*group_id] = true;
+	*cpu_dai_num = index;
+
+	return 0;
+}
+
+/* convert 1 -9 to '1' - '9' and A -F to 'A' - 'F' */
+static inline int int_to_ascii(int x)
+{
+	return x > 9 ? x + 87 : x + 48;
+}
+
+static int create_sdw_dailink(struct device *dev,
+			      struct snd_soc_dai_link *dai_links,
+			      int *be_id, int sdw_be_num, int sdw_cpu_dai_num,
+			      struct snd_soc_dai_link_component *cpus,
+			      const struct snd_soc_acpi_link_adr *link,
+			      int *cpus_id, bool *group_generated)
 {
-	struct snd_soc_dai_link_component *codec;
-	unsigned int part_id, link_id;
-	int i, j = 0, idx;
+	const struct snd_soc_acpi_link_adr *link_next;
+	struct snd_soc_dai_link_component *codecs;
+	int cpu_dai_id[SDW_MAX_DEVICES];
```

Comment:
that index max is wrong.
The SoundWire spec defines 11 devices PER LINK.
if you start dealing with mutilple links, then you need to multiply by the number of links.
And then we can have more than one DAI on a device.

In short, you need to come-up with a new max. e.g. SDW_MAX_CPU_DAIS 16
and check for overflows.